signac/contrib/project.py

```diff
@@ -1232,31 +1258,41 @@ def _get_statepoint(self, job_id, fn=None):
 
         """
         if not self._sp_cache:
+            # Triggers if no state points have been added to the cache, and all
+            # the values are None.
             self._read_cache()
         try:
-            if job_id in self._sp_cache:
-                return self._sp_cache[job_id]
-            else:
-                self._sp_cache_misses += 1
-                if (
-                    not self._sp_cache_warned
-                    and self._sp_cache_misses > self._sp_cache_miss_warning_threshold
-                ):
-                    logger.debug(
-                        "High number of state point cache misses. Consider "
-                        "to update cache with the Project.update_cache() method."
-                    )
-                    self._sp_cache_warned = True
-                statepoint = self._get_statepoint_from_workspace(job_id)
-        except KeyError as error:
+            # State point cache hit
+            return self._sp_cache[job_id]
+        except KeyError:
+            # State point cache missed
+            self._sp_cache_misses += 1
+            if (
+                not self._sp_cache_warned
+                and self._sp_cache_misses > self._sp_cache_miss_warning_threshold
+            ):
+                logger.debug(
```

Comment:
And does this error need to be tested as well?

Reply:
I see some lines aren't in the test coverage but I don't think there is very much that is actionable / worth doing. The changes here are just a re-work of the previous code and the crucial parts (fetching from cache, fetching from workspace, fetching from a file of state points) are all covered. https://codecov.io/gh/glotzerlab/signac/pull/239/src/signac/contrib/project.py

The lines I see that are missing coverage here are:
- Some error checking if the fallback to `read_statepoints` fails with an `OSError` (this is hard to test)
- The cache miss warning (this might be testable but it's just a logging message)

Reply:
Okay that makes sense!